frontend/utils/image.js

```diff
@@ -28,3 +29,88 @@ export const imageUpload = async (imageFile: File, params: ?Object): Promise<Obj
   const { download } = await sbp('chelonia/fileUpload', imageFile, { type: file.type, cipher: 'aes256gcm' }, params)
   return download
 }
+
+// Image compression
+
+export function supportsWebP (): Promise<boolean> {
+  // Uses a very small webP image to check if the browser supports 'image/webp' format.
+  // (reference: https://developers.google.com/speed/webp/faq#in_your_own_javascript)
+  const verySmallWebP = 'data:image/webp;base64,UklGRiIAAABXRUJQVlA4IBYAAAAwAQCdASoBAAEADsD+JaQAA3AAAAAA'
+  const img = new Image()
+
+  return new Promise(resolve => {
+    img.onload = () => { resolve(img.height > 0) }
+    img.onerror = (e) => { resolve(false) }
+    img.src = verySmallWebP
+  })
+}
+
+function loadImage (url): any {
+  const imgEl = new Image()
+
+  return new Promise((resolve) => {
+    imgEl.onload = () => { resolve(imgEl) }
+    imgEl.src = url
+  })
+}
+
+function generateImageBlobByCanvas ({
+  sourceImage,
+  resizingFactor,
+  quality,
+  compressToType
+}) {
+  const { naturalWidth, naturalHeight } = sourceImage
+  const canvasEl = document.createElement('canvas')
+  const c = canvasEl.getContext('2d')
+
+  canvasEl.width = naturalWidth * resizingFactor
+  canvasEl.height = naturalHeight * resizingFactor
+
+  c.drawImage(
+    sourceImage,
+    0,
+    0,
+    canvasEl.width,
+    canvasEl.height
+  )
+
+  return new Promise((resolve) => {
+    canvasEl.toBlob(blob => {
+      resolve(blob)
+    }, compressToType, quality)
+  })
+}
+
+export async function compressImage (imgUrl: string, sourceMimeType?: string): Promise<any> {
+  // Takes a source image url and generate a blob of the compressed image.
+
+  // According to the testing result, 0.8 is a good starting point for both resizingFactor and quality for .jpeg and .webp.
```

Comment:
which testing?

Reply:
IMO, this shouldn't be a 'resizing factor' anyhow, but it should resize images to some arbitrary maximum size. E.g., not the real suggestion, the max size could be 1024×768, and images are scaled not to exceed this. The dimensions would come from how the images are displayed in the chat.

Reply:
Just like I wrote in the desc, this codepen.. : https://codepen.io/freenaturalsoul/pen/vYoMWWv?editors=0100
I tried various large size images there. try it there.

> this shouldn't be a 'resizing factor' anyhow, but it should resize images to some arbitrary maximum size.

It doesn't seem like you are suggesting this without knowing what the goal of the issue..? The goal of the issue Greg wrote is to reduce the image **file-size** down below a certain value (400KB). Sacrificing the quality of the image size here is inevitable, at least when using `Canvas.toBlob()`. 

According to tests I've done, only changing the `quality` argument of that canvas API without reducing the image dimension along with it lead to the outcome image getting blurry earlier. That's why I chose to decrease the quality factor and reduce the dimension at the same time. Plus, that is something [this article](https://img.ly/blog/how-to-compress-an-image-before-uploading-it-in-javascript/) suggested too